src/extension/infer.rs

```diff
@@ -1133,6 +1160,502 @@ mod test {
+        )?;
+
+        let bb1 = make_block(
+            &mut hugr,
+            cfg,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: ExtensionSet::new(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(PRELUDE_ID, FunctionType::new(vec![NAT], oneway(NAT))),
+        )?;
+
+        // Mult
+        let bb2 = make_block(
+            &mut hugr,
+            cfg,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: ExtensionSet::new(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(PRELUDE_ID, FunctionType::new(vec![NAT], oneway(NAT))),
+        )?;
+
+        hugr.connect(entry, 0, bb0, 0)?;
+        hugr.connect(entry, 0, bb1, 0)?;
+        hugr.connect(bb0, 0, bb2, 0)?;
+        hugr.connect(bb1, 0, bb2, 0)?;
+        hugr.connect(bb2, 0, exit, 0)?;
+
+        hugr.infer_and_validate(&PRELUDE_REGISTRY)?;
+
+        Ok(())
+    }
+
+    ///       +-----------+
+    ///  +--->|   Entry   |
+    ///  |    +-----+-----+
+    ///  |          |
+    ///  |          V
+    ///  |    +------------+
+    ///  |    |    BB1     +---+
+    ///  |    +-----+------+   |
+    ///  |          |          |
+    ///  |          V          |
+    ///  |    +------------+   |
+    ///  +----+    BB2     |   |
+    ///       +------------+   |
+    ///                        |
+    ///       +------------+   |
+    ///       |    Exit    |<--+
+    ///       +------------+
+    fn make_cfg_loop_test(
+        entry_ext: ExtensionSet,
+        bb1_ext: ExtensionSet,
+        bb2_ext: ExtensionSet,
+    ) -> Result<Hugr, Box<dyn Error>> {
+        let hugr_delta = entry_ext.clone().union(&bb1_ext).union(&bb2_ext);
+
+        let mut hugr = Hugr::new(NodeType::open_extensions(ops::CFG {
+            signature: FunctionType::new(type_row![NAT], type_row![NAT])
+                .with_extension_delta(&hugr_delta),
+        }));
+
+        let root = hugr.root();
+
+        let exit = hugr.add_node_with_parent(
+            root,
+            NodeType::open_extensions(ops::BasicBlock::Exit {
+                cfg_outputs: type_row![NAT],
+            }),
+        )?;
+
+        let entry = {
+            let dfb: OpType = ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: entry_ext.clone(),
+            }
+            .into();
+
+            let entry = hugr.add_node_before(exit, NodeType::open_extensions(dfb))?;
+            let entry_in = hugr.add_node_with_parent(
```

Comment:
again this looks like `let (entry, entry_in, entry_out) = create_with_io(.....)` ?

Reply:
Here (and below) `create_with_io` is avoided because the dataflow parent needs to be added with `add_node_before` rather than just `add_node_with_parent`